lib/srv/forward/sshserver.go

```diff
@@ -457,38 +468,43 @@ func (s *Server) handleConnection(heartbeatContext context.Context, chans <-chan
 				return
 			}
 			go s.handleChannel(newChannel)
-		// If the heartbeats failed, close everything and cleanup.
-		case <-heartbeatContext.Done():
+		// If the server is closing (either the heartbeat failed or Close() was
+		// called, exit out of the connection handler loop.
+		case <-s.closeContext.Done():
 			return
 		}
 	}
 }
 
-func (s *Server) keepAliveLoop(cancel context.CancelFunc) {
+func (s *Server) keepAliveLoop() {
 	var missed int
 
-	// tick at 1/3 of the idle timeout duration
-	keepAliveTick := time.NewTicker(defaults.DefaultIdleConnectionDuration / 3)
-	defer keepAliveTick.Stop()
+	// Tick at 1/3 of the idle timeout duration.
+	tickerCh := time.NewTicker(defaults.DefaultIdleConnectionDuration / 3)
+	defer tickerCh.Stop()
 
 	for {
 		select {
-		case <-keepAliveTick.C:
-			// send a keep alive to the target node and the client to ensure both are alive.
+		case <-tickerCh.C:
+			// Send a keep alive to the target node and the client to ensure both are alive.
 			proxyToNodeOk := s.sendKeepAliveWithTimeout(s.remoteClient, defaults.ReadHeadersTimeout)
 			proxyToClientOk := s.sendKeepAliveWithTimeout(s.sconn, defaults.ReadHeadersTimeout)
 			if proxyToNodeOk && proxyToClientOk {
 				missed = 0
 				continue
 			}
 
-			// if we miss 3 in a row the connections dead, call cancel and cleanup
+			// If 3 keep alives are missed, the connection is dead, call cancel and cleanup.
 			missed += 1
 			if missed == 3 {
-				s.log.Infof("Missed %v keep alive messages, closing connection", missed)
-				cancel()
+				s.log.Infof("Missed %v keep alive messages, closing connection.", missed)
+				s.closeCancel()
```

Comment:
why call `closeCancel` here and not `Close`? this call seems to be very low level?

Reply:
Close() will be called twice then once here then once again in `handleConnection`.